### DISCUSSION TITLE
Extracting hypno as csv?

### DISCUSSION THREAD

**Discussion by PhD-GOAT** [raphaelvallat/yasa#65] [Q&A] [answered]
[Sleep stages (aka hypnogram, .csv format, 3KB)](https://drive.google.com/uc?id=1s-XXBIXt0YKsbihMa1imBo4oftpb1Ds_), in 30-seconds epoch. The sleep stages are encoded as follow: 0 = Wake, 1 = N1, 2 = N2, 3 = N3, and 4 = REM.

In your quick start, hypno is a one-dimensional arrangement. But normally, clinical data doesn't seem to be one-dimensional, so how do you convert str arrays?

**Comment by raphaelvallat** [comment, marked as answer]
Hi @PhD-GOAT I don't understand what you mean by "one-dimensional", please clarify

**Reply by PhD-GOAT** [reply]
![image](https://user-images.githubusercontent.com/93689515/167516239-ae260d5a-6f2d-45b0-803e-42bbe49fdc02.png)

If you open your csv file, it's a one-dimensional column vector as shown in the picture.

So the problem is, I want to extract the actual hypno csv instead of your example hypno

**Reply by remrama** [reply]
yo. @PhD-GOAT. that's a csv file with 1 column. load it into python with `pd.read_csv` or `np.loadtxt` or whatever else.

**Reply by PhD-GOAT** [reply]
yeah, i just tried

![image](https://user-images.githubusercontent.com/93689515/167519468-830d3705-95c3-4f0f-8f14-c185262b1b81.png)

If just read the data frame, it comes out in multi-dimensions. What do you do? If you want to use yasa api, you need to be one-dimensional to do it.

**Reply by remrama** [reply]
Unfortunately I don't know what you did that resulted in that screenshot...
@PhD-GOAT **please include the code you are running** to try and load this file in

**Reply by PhD-GOAT** [reply]
`hypno = 'C:/Users/SC4001EC-Hypnogram.edf';
df = pd.read_csv(hypno);
print(df)`

The output is shown above.

**Reply by remrama** [reply]
Thanks @PhD-GOAT that makes this way easier!!

You are trying to read the `.edf` file into pandas as a csv. You've been referring to the `.csv` file this whole time, so I presume this was just a simple mix-up. It should work if you have anything like
```
hypno = 'C:/Users/SC4001EC-Hypnogram.csv' # or whatever your CSV filename is
df = pd.read_csv(hypno)
print(df)
```

**Reply by PhD-GOAT** [reply]
How can I do that command when the edf file can't be changed to csv from the first place? I don't understand.

**Reply by remrama** [reply]
Your original post includes a link _to a csv file_, so I was under the impression that's what you were trying to load in. You don't need to convert the edf to csv. Forget about the edf and download the csv that you linked to in your original post. Name that accordingly and load it in.

**Reply by PhD-GOAT** [reply]
https://physionet.org/content/sleep-edfx/1.0.0/

I referred to this site, but I don't think there is a separate csv file. First of all, thank you.

**Reply by remrama** [reply]
Oh okay. I think I'm having a moment of clarity.

The hypnogram from the **physionet** dataset comes in edf format. The hypnogram from **yasa** comes in csv format. So you are able to load in the yasa hypnogram using pandas, but you're struggling with the physionet hypnogram. Well, they're different filetypes so you can't treat them interchangeably. You will have to load the physionet hypnogram using [MNE](https://mne.tools/stable/index.html) and explore that file to extract the hypnograms. Notably, I don't think this is the place to discuss that further.

Based on how this and many of your other threads are ending, I am now realizing that your struggles are with the physionet dataset specifically and probably don't relate to yasa in any specific way. I mean, the issues are not about yasa's ability to handle the data, or how yasa works, it's just about the structure/format of the physionet dataset and how to read it. It's not as simple as running just _any_ edf file through the tutorial scripts and expecting things to run smoothly.

In my opinion the yasa discussion threads are probably not the best place to discuss the details of the physionet dataset (again, unless there are some yasa-specific issues or something, but I don't think that is or would be the case). I understand why it seems relevant and how the questions are arising, but ultimately the only connection is that it's sleep data. @PhD-GOAT **my suggestion is to dedicate some time to understanding the ins-and-outs of the physionet dataset before trying to analyze it in detail.** That may or may not involve finding another resource for asking questions about how to read it into python. I hope that helps.

**Reply by PhD-GOAT** [reply]
I agree. So I wonder if I should change the API of Yasa by myself or modify the data if it is a physical data problem. If it is the former, I think there will be a functional problem because I changed it. Don't you think so?

**Reply by remrama** [reply]
Great! Now we're talking!

 You do not need to mess with yasa in any way. It's not going to solve any of your issues. Again, this isn't related to how yasa handles data. Instead, you need to work on extracting the data you want from physionet before even considering how to process it with yasa. Don't even load in yasa until you have the data you want from physionet. For example, you're final script might involve a top section that is purely dedicated to extracting the relevant physionet data and getting it to look the way yasa wants it, only to _then_ process it below with yasa.

So you need to play with physionet in python before loading yasa. I would suggest looking at [this page](https://mne.tools/stable/auto_tutorials/clinical/60_sleep.html#sphx-glr-auto-tutorials-clinical-60-sleep-py) in detail. That tutorial classifies sleep stages using just MNE and sklearn, and it uses the physionet dataset! So you should see how they extract hypnograms in that script.

Now yasa will likely outperform that tutorial, so you just want to use that to understand physionet data in python and then take what you want to run your own yasa script. I'd still suggest working all the way through that one before yanking from it.
